haystack/utils/jinja2.py

```diff
@@ -0,0 +1,39 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Optional
+
+from jinja2 import Environment, nodes
+
+
+def extract_declared_variables(template_str: str, env: Optional[Environment] = None) -> set:
+    """
+    Extract declared variables from a Jinja2 template string.
+
+    Args:
+        template_str (str): The Jinja2 template string to analyze.
+        env (Environment, optional): The Jinja2 Environment. Defaults to None.
+
+    Returns:
+        A list of variable names used in the template.
```

Comment:
Let's make sure to use the docstring formatting consistent with the rest of the library so

```suggestion
    Extract declared variables from a Jinja2 template string.

    :param template_str: The Jinja2 template string to analyze.
    :param env: The Jinja2 Environment. Defaults to None.

    :returns:
        A set of variable names used in the template.
```

Also update the returns docstring to reflect we are returning a set not a list